Sources/DeepLearning/Trainer.swift

```diff
@@ -0,0 +1,47 @@
+// Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// SupervisedTrainer implements a standard training loop for a model.
+public struct SupervisedTrainer<Model, Opt: Optimizer, Loss: Differentiable>
+    where Opt.Model == Model,
+          Loss: FloatingPoint,
+          Loss == Loss.CotangentVector {
+    // TODO: Convert to protocol to take advantage of argument labels to disambiguate.
+    typealias LossFn = @differentiable (Model.Output, Model.Output) -> Loss
+
+    var model: Model
+    var optimizer: Opt
+    var lossFn: LossFn
+
+    init(model: Model, with optimizer: Opt, loss: @escaping LossFn) {
+        self.model = model
+        self.optimizer = optimizer
+        self.lossFn = loss
+    }
+
+    mutating func fit(x: Model.Input, y: Model.Output, forSteps numSteps: Int) {
```

Comment:
It would make the current "copying model at the training initializer" semantics (which is confusing) go away if you make model an inout argument taken in the `fit` function.

Reply:
Right; I didn't particularly like having the model be copied into the trainer (because the local variable in the user's program is untrained, but would be very easy to use by accident.

Does it instead make sense for this to be a freestanding function (or better yet a type method to avoid polluting the global namespace)?

Reply:
I think defining this as a freestanding function is actually a good idea.

Reply:
Refactored; resolving now.